core/src/main/scala/kafka/tools/DumpLogSegments.scala

```diff
@@ -150,6 +164,70 @@ object DumpLogSegments {
     }
   }
 
+  private def dumpTimeIndex(file: File,
+                            indexSanityOnly: Boolean,
+                            verifyOnly: Boolean,
+                            misMatchesForIndexFilesMap: mutable.HashMap[String, List[(Long, Long)]],
+                            maxMessageSize: Int) {
+    val startOffset = file.getName().split("\\.")(0).toLong
+    val logFile = new File(file.getAbsoluteFile.getParent, file.getName.split("\\.")(0) + Log.LogFileSuffix)
+    val messageSet = new FileMessageSet(logFile, false)
+    val indexFile = new File(file.getAbsoluteFile.getParent, file.getName.split("\\.")(0) + Log.IndexFileSuffix)
+    val index = new OffsetIndex(indexFile, baseOffset = startOffset)
+    val timeIndex = new TimeIndex(file, baseOffset = startOffset)
+
+    //Check that index passes sanityCheck, this is the check that determines if indexes will be rebuilt on startup or not.
+    if (indexSanityOnly) {
+      timeIndex.sanityCheck
+      println(s"$file passed sanity check.")
+      return
+    }
+
+    var prevTimestamp = Message.NoTimestamp
+    for(i <- 0 until timeIndex.entries) {
+      val entry = timeIndex.entry(i)
+      val position = index.lookup(entry.offset).position
+      var partialFileMessageSet: FileMessageSet = messageSet.read(position, maxMessageSize)
+      val shallowIter = partialFileMessageSet.iterator
+      var maxTimestamp = Message.NoTimestamp
+      var checkedOffset = -1L
+      // We stop scan when on of the following is not met:
+      // 1. All the messages are checked
+      // 2. The max timestamp checked is greater or equals to the indexed timestamp
+      // 3. When the offset checked is greater than the next indexed offset.
+      var wrapperMessageOpt = if (shallowIter.hasNext) Some(shallowIter.next) else None
+      do {
+        wrapperMessageOpt match {
+          case Some(wrapperMessage) =>
+            val deepIter = getIterator(wrapperMessage, isDeepIteration = true)
+            for (messageAndOffset <- deepIter)
+              maxTimestamp = math.max(maxTimestamp, messageAndOffset.message.timestamp)
+            checkedOffset = wrapperMessage.offset
+            wrapperMessageOpt = if (shallowIter.hasNext) Some(shallowIter.next) else None
+          case _ =>
+        }
+      } while (wrapperMessageOpt.isDefined && maxTimestamp < entry.timestamp &&
```

Comment:
The wrapperMessageOpt will be updated in line 206. We need to do this because if user is producing uncompressed messages, a "batch" would actually be multiple messages. The first message in the batch may not always have the the largest timestamp. So we need to scan a little. We should be able to find the indexed timestamp before the next indexed offset in the time index or the log end, whichever comes first.